packages/remix-express/server.ts

```diff
@@ -38,14 +50,52 @@ export type RequestHandler = (
 export function createRequestHandler({
   build,
   getLoadContext,
+  serverMiddleware,
   mode = process.env.NODE_ENV,
 }: {
   build: ServerBuild;
   getLoadContext?: GetLoadContextFunction;
+  serverMiddleware?: ServerMiddlewareFunction;
   mode?: string;
 }): RequestHandler {
   let handleRequest = createRemixRequestHandler(build, mode);
 
+  if (build.future.unstable_middleware && serverMiddleware) {
+    return async (
+      req: express.Request,
+      res: express.Response,
+      next: express.NextFunction
+    ) => {
+      let response: NodeResponse | undefined;
+      let context = createMiddlewareStore();
+      let callRemix = async () => {
+        let request = createRemixRequest(req, res);
+        response = (await handleRequest(
+          request,
+          undefined,
+          context
+        )) as NodeResponse;
+        return response;
+      };
+      context.next = callRemix;
+
+      try {
+        await serverMiddleware({ request: req, response: res, context });
+        if (!response) {
+          // User never called next(), so doesn't need to do any post-processing
+          response = await callRemix();
+        }
+        if (!res.headersSent) {
+          await sendRemixResponse(res, response);
+        }
```

Comment:
A few things going on here:

* If the user never calls `context.next()` we automatically call the rest of the pipeline.  You cannot short circuit by skipping the call.  The only way to short circuit is to `throw`
* If the user actually responds via the express `Response` in the middleware (and thus `res.headersSent === true`) then we skip `sendRemixResponse()`.  This allows them to take full control if they want to.